server/cluster/unsafe_recovery_controller.go

```diff
@@ -485,6 +486,7 @@ func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
 		output.Info = "Unsafe recovery enters exit force leader stage"
 		if u.err != nil {
 			output.Details = append(output.Details, fmt.Sprintf("triggered by error: %v", u.err.Error()))
+			exitForceLeaderDueToFailure = true
```

Comment:
how about not calling `changeStage(exitForceLeader)` in `handleErr`, and set stage with `exitForceLeader`
``` go
var stage Stage
if u.err == nil {
  stage = u.stage
} else {
  stage = exitForceLeader
}

reCheck := false
for {
	switch stage {
	case collectReport:
		fallthrough
	case tombstoneTiFlashLearner:
		.....
````

Reply:
Updated, it looks a bit messy though. PTAL

Reply:
That's really messy. Why not just not calling `changeStage(exitForceLeader)` in `handleErr`, and set stage with `exitForceLeader`. We need to collect the report again to make sure the TiKVs are notified.

Reply:
Updated, per our discussion offline.